### PR TITLE
Make sanitizeURL deliver more consistent results

### PR DESCRIPTION
**Pull request by cyberpower678**
The sanitizeURL function is a publicly accessible function, it should
deliver some consistency when sanitizing multiple instances of the same
URL.  Will be used by IABot in the future.

### CODE REVIEW

File: src/CheckIfDead.php [hunk 1]
```diff
@@ -412,27 +417,51 @@ public function sanitizeURL( $url ) {
...
-							explode( '/',
-								substr(
-									urldecode( $parts['path'] ), 1
-								)
-							)
+				array_map( "rawurlencode",
+					explode( '/',
+						substr(
+							urldecode( $parts['path'] ), 1
 						)
```
Comment (cyberpower678) [inline src/CheckIfDead.php]: I restored the indenting back to how it's supposed to be here.  No functional change here.

File: src/CheckIfDead.php [hunk 2]
```diff
@@ -445,11 +474,11 @@ public function sanitizeURL( $url ) {
 				// Make sure we don't inadvertently encode the first instance of "="
 				// Otherwise we break the query.
 				$parts['query'][$index] = implode( '=',
-											array_map( "urlencode",
-												array_map( "urldecode",
-														   explode( '=', $parts['query'][$index], 2 )
-												)
-											)
+					array_map( "urlencode",
```
Comment (cyberpower678) [inline src/CheckIfDead.php]: Also fixed the indenting here.

File: src/CheckIfDead.php [hunk 3]
```diff
@@ -412,27 +417,51 @@ public function sanitizeURL( $url ) {
...
+						}
+					case 21:
+						if ( isset( $parts['scheme'] ) &&
+							strtolower( $parts['scheme'] ) == "ftp"
+						) {
+							break;
+						}
+					default:
+						$url .= ":" . $parts['port'];
+				}
```
Comment (cyberpower678) [inline src/CheckIfDead.php]: This is a functional change.  Purges ports from the URL if it's HTTP :80, HTTPS  :443, or FTP :21.

File: src/CheckIfDead.php [hunk 4]
```diff
@@ -412,27 +417,51 @@ public function sanitizeURL( $url ) {
 		if ( isset( $parts['host'] ) ) {
 			// Properly encode the host.  It can't be UTF-8.
 			// See https://en.wikipedia.org/wiki/Internationalized_domain_name.
+			// Hosts are case insensitive, might as well make them lowercase too.
 			if ( function_exists( 'idn_to_ascii' ) ) {
-				$url .= idn_to_ascii( $parts['host'] );
+				$url .= strtolower( idn_to_ascii( $parts['host'] ) );
 			} else {
-				$url .= $parts['host'];
+				$url .= strtolower( $parts['host'] );
```
Comment (cyberpower678) [inline src/CheckIfDead.php]: This adds consistency when called by the external application.

File: src/CheckIfDead.php [hunk 5]
```diff
@@ -388,15 +394,14 @@ public function sanitizeURL( $url ) {
...
 		// Break up the URL first
 		$parts = $this->parseURL( $url );
-
 		// In case the protocol is missing, assume it goes to HTTPS
+		// Schemes are case insensitive, might as well make them lowercase
 		if ( !isset( $parts['scheme'] ) ) {
 			$url = "https";
 		} else {
-			$url = $parts['scheme'];
+			$url = strtolower( $parts['scheme'] );
```
Comment (cyberpower678) [inline src/CheckIfDead.php]: This adds some consistency when called by the external application.

File: src/CheckIfDead.php [hunk 6]
```diff
@@ -298,6 +297,7 @@ protected function processCurlResults( $curlInfo, $full = false ) {
 		if ( $httpCode >= 400 && $httpCode < 600 ) {
 			if ( $full ) {
 				$this->errors[$curlInfo['rawurl']] = "RESPONSE CODE: $httpCode";
+
```
Comment (Niharika29) [inline src/CheckIfDead.php]: Why all these unrelated blank lines?

**Comment by cyberpower678** [comment]
I'm not sure exactly, PHPStorm is simply applying the PHP Code style settings you gave me.  It would be nice if we could simply merge the code style changes and be done with it.

File: src/CheckIfDead.php [hunk 7]
```diff
@@ -412,27 +417,51 @@ public function sanitizeURL( $url ) {
...
+				// Remove port numbers if they are the default.
+				switch ( $parts['port'] ) {
+					case 80:
+						if ( isset( $parts['scheme'] ) &&
+							strtolower( $parts['scheme'] ) == "http"
+						) {
+							break;
+						}
+					case 443:
+						if ( !isset( $parts['scheme'] ) ||
```
Comment (Niharika29) [inline src/CheckIfDead.php]: You mean if ( isset(...

File: src/CheckIfDead.php (hunk 3 above)
Comment (Niharika29) [inline src/CheckIfDead.php]: It took me a while to understand what this code is doing. I think there's a better way to write this. How about the following?
```
$scheme = strtolower( $parts['scheme'] );
if ( isset( $parts['port'] && isset( $parts['scheme'] ) ) {
	// If port numbers are not default for the scheme, add them
	if ( !( $parts['port'] == 80 && $scheme == 'http' ) ||
		!( $parts['port'] == 443 && $scheme == 'https' ) ||
		!( $parts['port'] == 21 && $scheme == 'ftp' )
	) {
		$url .= ":" . $parts['port'];
	}
} elseif ( isset( $parts['port'] ) ) {
	// Add port number if scheme is unknown
	$url .= ":" . $parts['port'];
}
```

**Comment by cyberpower678** [comment]
Your code assumes the scheme is set.  Curl can't handle protocol relative URLs.  In the event the class is given one, it assumes https, also hence the !isset.

**Comment by cyberpower678** [comment]
I also considered using an if statement but that would have resulted in a nasty logic expression.  That's why I decided to go with switch.

File: src/CheckIfDead.php (hunk 3 above)
Comment (cyberpower678) [inline src/CheckIfDead.php]: That code is buggy.  It assumes $parts['scheme'] is already set.

Comment (Niharika29) [inline src/CheckIfDead.php]: You can move the $scheme initiation inside of the first if.

Comment (cyberpower678) [inline src/CheckIfDead.php]: I also don't know what 
`elseif ( isset( $parts['port'] ) ) {
	// Add port number if scheme is unknown
	$url .= ":" . $parts['port'];
}`

is supposed to be.  We can't curl a scheme less URL.

**Comment by Niharika29** [comment]
You're making a new PR?
